plugins/modules/vmware_guest_disk.py

```diff
@@ -642,31 +693,39 @@ def ensure_disks(self, vm_obj=None):
-                                disk_spec.device.backing.fileName = "[%s] %s/%s_%s_%s_%s.vmdk" % (disk['datastore'].name,
-                                                                                                  self.vm.name,
-                                                                                                  self.vm.name,
-                                                                                                  device.key,
-                                                                                                  str(disk['disk_unit_number']),
-                                                                                                  str(randint(1, 10000)))
-                        elif disk['filename'] is not None:
-                            disk_spec.device.backing.fileName = disk['filename']
-                        disk_spec = self.get_ioandshares_diskconfig(disk_spec, disk)
+                        else:
+                            if disk['filename'] is None:
+                                if disk['datastore_cluster'] is not None:
+                                    datastore_name = self.get_recommended_datastore(datastore_cluster_obj=disk['datastore_cluster'], disk_spec_obj=disk_spec)
+                                    disk['datastore'] = find_obj(self.content, [vim.Datastore], datastore_name)
+
+                                disk_spec.fileOperation = vim.vm.device.VirtualDeviceSpec.FileOperation.create
+                                disk_spec.device.capacityInKB = disk['size']
+                                # Set backing filename when datastore is configured and not the same as VM datastore
+                                # If datastore is not configured or backing filename is not set, default is VM datastore
+                                if disk['datastore'] is not None and disk['datastore'].name != vm_files_datastore:
+                                    disk_spec.device.backing.datastore = disk['datastore']
+                                    disk_spec.device.backing.fileName = "[%s] %s/%s_%s_%s_%s.vmdk" % (disk['datastore'].name,
+                                                                                                      self.vm.name,
+                                                                                                      self.vm.name,
+                                                                                                      device.key,
+                                                                                                      str(disk['disk_unit_number']),
+                                                                                                      str(randint(1, 10000)))
+                            elif disk['filename'] is not None:
+                                disk_spec.device.backing.fileName = disk['filename']
+                                disk_spec = self.get_ioandshares_diskconfig(disk_spec, disk)
```

Comment:
This line `disk_spec = self.get_ioandshares_diskconfig(disk_spec, disk)` is not only for 'elif'.

Reply:
I looked at this and this is not something I edited. I only wrapped the pre-exising logic around an if statement in 703-707 and move it to an else block. I was trying to reduce the changes to others code. I can update to a "else:" if you think that will work better?

Reply:
Sure, I saw the logic is as below, you may not notice the indent of that line, it's not inside the "elif" statement.
```
if disk['filename'] is None:
...                          
elif disk['filename'] is not None:
    disk_spec.device.backing.fileName = disk['filename']
disk_spec = self.get_ioandshares_diskconfig(disk_spec, disk)
```

Reply:
Thanks @Tomorrow9! I see it now, I updated the indent and pushed it just now.

Reply:
Thank you.